### PR TITLE
Reduce excessive logging in debug.log and perf.log

### PR DESCRIPTION
**Pull request by reaperrr**
Following https://github.com/OpenRA/OpenRA/issues/19029#issuecomment-803683420, this PR is meant to complement #19264 and/or other solutions to hopefully solve #19260 for good.

While having as much information as possible available to debug any crashes is generally valuable, the amount of of logging we do seems excessive for the purpose it's mean to serve.

Logging the `--Tick` in `debug.log` seems mostly pointless, as the few kinds of potential crash sources getting logged in debug.log either happen at/before game start, or can include the world tick in their log entry.

Meanwhile, the perf.log of players rarely shows us devs anything we don't already know (except infinite loops when their cost exceeds the threshold, but that's a poor excuse to drag down performance for players).

Additionally, in both cases developers can still set their preferred value in settings.yaml and run a provided replay to get the needed information, if necessary.

### CODE REVIEW

**Comment by dnqbob** [comment]
I think CastleJing will like this idea and he planned to do similar change to TA-engine

**Comment by dnqbob** [comment]
For me I want both #19264 and this PR.

**Comment by pchote** [comment]
Making this opt-in is a good idea, but changing the default to 10ms makes the feature essentially useless. The idea of this logging was to highlight the intermediate cases that aren't bad enough to cause an obvious lag spike on their own, but when combined with other intermediate cases in the same tick still completely blow the frame budget. We already know and mostly can't fix the 10ms+ cases, so don't care about those!

**Comment by dnqbob** [comment]
Maybe use a loop queue to record last few ticks?

**Comment by reaperrr** [comment]
> Making this opt-in is a good idea, but changing the default to 10ms makes the feature essentially useless. The idea of this logging was to highlight the intermediate cases that aren't bad enough to cause an obvious lag spike on their own, but when combined with other intermediate cases in the same tick still completely blow the frame budget. We already know and mostly can't fix the 10ms+ cases, so don't care about those!

Point taken. But making the perf-logging opt-in as well would be fine? Or at least making it opt-out?

**Comment by pchote** [comment]
IMO we should treat it the same way we do `SyncCheckUnsyncedCode`: disabled by default, with a checkbox hidden in the Developer Settings menu.

**Comment by reaperrr** [comment]
OK, though that would make us circle back to the lack of space in the Advanced settings, especially once #19242 is merged as well.

Would squeezing the debug checkboxes closer together (vertically) be acceptable until the menu is restructured?

**Comment by pchote** [comment]
#19242 won't work as a debug option (see my comment there), which solves the issue here 🙂

**Comment by reaperrr** [comment]
Before I update this, some actual performance numbers.
I measured by logging the average of the last 500 ticks (via PerfHistory) at ticks 501 and 1001.
Also did 5 runs each with and without logging enabled, to reduce impact of fluctuations from CPU clockspeeds or background tasks.

RA shellmap with perf logging enabled, like on bleed:
```
Run 1:
--Tick: 501, 2,649339393939394 ms
--Tick: 1001, 2,6562929292929276 ms

Run 2:
--Tick: 501, 2,6692050505050493 ms
--Tick: 1001, 2,664829292929293 ms

Run 3:
--Tick: 501, 2,6916747474747464 ms
--Tick: 1001, 2,684061616161616 ms

Run 4:
--Tick: 501, 2,670688888888888 ms
--Tick: 1001, 2,673647474747474 ms

Run 5:
--Tick: 501, 2,651198989898989 ms
--Tick: 1001, 2,673625252525253 ms
```

With perf logging (including the GetTimestamp calls) disabled:
```
Run 1:
--Tick: 501, 2,4912707070707074 ms
--Tick: 1001, 2,470488888888889 ms

Run 2:
--Tick: 501, 2,424936363636364 ms
--Tick: 1001, 2,4316414141414135 ms

Run 3:
--Tick: 501, 2,584616161616162 ms
--Tick: 1001, 2,498614141414141 ms

Run 4:
--Tick: 501, 2,42079191919192 ms
--Tick: 1001, 2,4808181818181816 ms

Run 5:
--Tick: 501, 2,5107303030303028 ms
--Tick: 1001, 2,544620202020202 ms
```

Looks like about 5-7% less tick time on the RA shellmap just from disabling perf logging, and we're talking about a Ryzen 3600 here, on slower systems + larger games with significantly more perf logging we'd likely see a 2-digit performance impact.

Of course, #19264 will likely considerably reduce the impact of this PR, but at minimum we'll still save some `Stopwatch.GetTimestamp` calls.

**Comment by reaperrr** [comment]
> which solves the issue here 🙂

Technically, I'd need 2 checkboxes, unless we want to merge the options into a generic "Additional debug/perf logging" option.
For now I only exposed the perf logging option, since I consider that one more important.

**Comment by dnqbob** [comment]
I will suggest some mods that already slowed by trait, map and AI use log record fix PR

File: OpenRA.Game/Game.cs
Comment (pchote) [inline OpenRA.Game/Game.cs]: IMO we should just delete this entirely.

File: OpenRA.Mods.Common/Traits/Health.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Health.cs]: And these too. Do we have any other tick-based debug output?

File: OpenRA.Mods.Common/Traits/Render/WithCargoPipsDecoration.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithCargoPipsDecoration.cs]: This would ideally be caught by a lint test. Nobody actually reads debug.log afaik so we could delete this too and leave it up to modders to realize that the game is falling back to the default pip sequence to identify that something is wrong.

**Review by pchote — COMMENTED** [review]
We can remove the need for two options by removing the second set of logging. Most of these were added years ago to help solve specific problems, then nobody got around to cleaning them up before now.

File: OpenRA.Game/Game.cs
Comment (pchote) [reply OpenRA.Game/Game.cs]: On second thought, this is actually an interesting number to extract and plot so how about we move this to perf.log and the perf debug setting?

File: OpenRA.Mods.Common/Traits/Health.cs
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/Health.cs]: > Do we have any other tick-based debug output?

Not really. I guess I'll just remove it.

File: OpenRA.Game/Game.cs
Comment (pchote) [reply OpenRA.Game/Game.cs]: On third thought, this is just logging the tick numbers without any performance information. Ignore my last comment and lets just delete this!

**Comment by reaperrr** [comment]
Updated.

Also pulled out `var`s for the debug option to make things a bit more readable (maybe saving a microsecond per lookup as well).

File: OpenRA.Mods.Common/Traits/Render/WithCargoPipsDecoration.cs
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithCargoPipsDecoration.cs:72]: You forgot to undo removing this as the if statement above seems to lead nowhere?

Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/Render/WithCargoPipsDecoration.cs:72]: Derp. Well, removing this was intentional, but of course the if can't stay like that, good catch.

File: OpenRA.Mods.Common/Traits/Render/WithCargoPipsDecoration.cs
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Render/WithCargoPipsDecoration.cs]: I haven't used TryGetValue before, so please check if it's fine to use it this way.

File: OpenRA.Mods.Common/Traits/World/ValidateOrder.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/ValidateOrder.cs]: Style nit: can we please keep the formatting arguments in order? (renumber them so tick is {0})

File: OpenRA.Mods.Common/Traits/Render/WithCargoPipsDecoration.cs
Comment (pchote) [reply OpenRA.Mods.Common/Traits/Render/WithCargoPipsDecoration.cs]: Almost but not quite!

The problem is that `TryGetValue` needs to set a placeholder value (in this case null) to the out value when the key isn't found. Returning the now-null sequence will cause a crash when trying to render the pip.

The best practice is to to put the `TryGetValue` in an `if` and either directly return the value on success (if that fits the usecase) or to override the invalid out value with a suitable placeholder.

In this case, because we already have to check the CustomPipType, the first case works best:
```csharp
if (n < pi.Weight)
{
	if (pi.CustomPipType != null && Info.CustomPipSequences.TryGetValue(pi.CustomPipType, out var sequence))
		return sequence;

	return Info.FullSequence;
}
```

**Comment by reaperrr** [comment]
Updated.

**Comment by teinarss** [comment]
We have the same logic in 3 places and it also makes the code a bit harder to read, one idea is to move the logic to a dedicated class see https://github.com/teinarss/OpenRA/commit/cd363e35a08befb4c184898303f026fb1d2f9a12 for a example how we could do it. We could also let the ShouldLog method be private and let Log call it.

**Comment by reaperrr** [comment]
Updated (and rebased).

Compared to @teinarss code, I merged the ShouldLog() method into Log(), since the actual logging must happen before `start` is reset so we can't do this in separate methods (unless we cache it, but I felt just merging the methods is a simpler solution).

File: OpenRA.Game/TraitDictionary.cs
Comment (teinarss) [inline OpenRA.Game/TraitDictionary.cs]: no need for this :)

File: OpenRA.Game/TraitDictionary.cs
Comment (pchote) [inline OpenRA.Game/TraitDictionary.cs]: Style nit: can be made readonly, and grouped with the two above.

File: OpenRA.Game/TraitDictionary.cs
Comment (pchote) [inline OpenRA.Game/TraitDictionary.cs]: Its not clear from the method name that this also restarts the timer, which isn't great IMO.

How about changing this to `RestartAndLogLongTick`? Its a bit of a mouthful so please substitute with something better if you can think of it.

File: OpenRA.Game/TraitDictionary.cs
Comment (pchote) [inline OpenRA.Game/TraitDictionary.cs]: Can we please put this in its own file in the OpenRA.Game/Support directory (next to PerfTimer etc)?

TraitDictionary isn't a good home for this considering the other places that also use it - although this is less of an issue if we change those places to use the TraitDictionary copy.

File: OpenRA.Game/TraitDictionary.cs
Comment (pchote) [inline OpenRA.Game/TraitDictionary.cs]: This creates an awkward UI situation: `TraitDict` creates its `SimulationPerfLogging` when loading and then reuses it for the duration of the game. This means that this value won't be updated if the player updates the option in the settings menu. Right now RunActivity and DoTimed create new objects each call, so they will update.

The simplest solution would be to update this value from `Start()`. We could cache `Game.Settings.Debug` to only incur one level of pointer chasing.

File: OpenRA.Game/Traits/ActivityUtils.cs
Comment (pchote) [inline OpenRA.Game/Traits/ActivityUtils.cs]: IMO its bad practice to reference Stopwatch.GetTimestamp here now - we might change the internal details of `SimulationPerfLogging` in the future, and probably won't know to update these comments.

I would personally replace this first comment with
```csharp
// PERF: This is a hot path and must run with minimal added overhead.
// If there are no activities we can bail straight away and save ourselves the overhead of setting up the perf logging.
```
and then delete the second comment completely.

What do you think?

File: OpenRA.Game/WorldUtils.cs
Comment (pchote) [inline OpenRA.Game/WorldUtils.cs]: Likewise here.

File: mods/cnc/chrome/settings-advanced.yaml
Comment (pchote) [inline mods/cnc/chrome/settings-advanced.yaml]: Style nit: these checkboxes generally use title case, so `Enable Additional Performance Logging`, or maybe `Enable Tick Performance Logging`?

File: OpenRA.Game/Traits/ActivityUtils.cs
Comment (pchote) [inline OpenRA.Game/Traits/ActivityUtils.cs]: Would there be any value in making `self.World.TraitDict.perfLogging` public and then reusing that here?

File: OpenRA.Game/WorldUtils.cs
Comment (pchote) [inline OpenRA.Game/WorldUtils.cs]: Likewise here, although it would need to be passed as an argument into the method.

File: OpenRA.Game/TraitDictionary.cs
Comment (pchote) [inline OpenRA.Game/TraitDictionary.cs]: You missed a few other style issues in here 🙂 

The elses are redundant (since the ifs return/throw) and `default(T)` can now simplify to just `default`.

File: OpenRA.Game/TraitDictionary.cs
Comment (reaperrr) [reply OpenRA.Game/TraitDictionary.cs]: > We could cache Game.Settings.Debug to only incur one level of pointer chasing.

Not sure how to do that, so I skipped that part for now.

File: OpenRA.Game/Traits/ActivityUtils.cs
Comment (reaperrr) [reply OpenRA.Game/Traits/ActivityUtils.cs]: No idea.

I went with moving it to its own file, so I guess it doesn't matter anymore (unless we change it back).

File: OpenRA.Game/TraitDictionary.cs
Comment (pchote) [reply OpenRA.Game/TraitDictionary.cs]: The same as this line, basically:

```csharp
DebugSettings settings = Game.Settings.Debug;
```

**Comment by reaperrr** [comment]
Updated. Used the opportunity to rename the logger to better match the other Perf* measure helpers.

**Review by pchote — DISMISSED** [review]

File: OpenRA.Game/Support/PerfTickLogger.cs
Comment (teinarss) [inline OpenRA.Game/Support/PerfTickLogger.cs]: You can remove all expect `using System.Diagnostics;`

File: OpenRA.Game/Support/PerfTickLogger.cs
Comment (teinarss) [inline OpenRA.Game/Support/PerfTickLogger.cs]: Maybe make it sealed

File: OpenRA.Game/Support/PerfTickLogger.cs
Comment (teinarss) [inline OpenRA.Game/Support/PerfTickLogger.cs]: readonly

File: OpenRA.Game/Traits/ActivityUtils.cs
Comment (teinarss) [inline OpenRA.Game/Traits/ActivityUtils.cs:21]: In this file you can remove 

```
using System.Diagnostics;
using OpenRA.Activities;

```

File: OpenRA.Game/TraitDictionary.cs
Comment (teinarss) [inline OpenRA.Game/TraitDictionary.cs:148]: In this file you can remove 

```
using System.Diagnostics;
```

File: OpenRA.Game/WorldUtils.cs
Comment (teinarss) [inline OpenRA.Game/WorldUtils.cs:68]: In this file you can remove 

```
using System.Diagnostics;
```

**Comment by reaperrr** [comment]
Updated.

**Review by teinarss — APPROVED** [review]

**Comment by teinarss** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7cf3520927d17c583136b9fd22e53988705e0db9...3269d9263953f5dc1ad2ec124dfdfe30362727c6

**Pull request merged**
